sphinx_doc/tutorials/docs/writing_new_costmap2d_plugin.rst

```diff
@@ -0,0 +1,240 @@
+.. _writing_new_costmap2d_plugin:
+
+Writing a new Costmap2D plugin
+******************************
+
+- `Overview`_
+- `Requirements`_
+- `Tutorial Steps`_
+
+.. image:: images/Writing_new_Costmap2D_plugin/gradient_layer_preview.gif
+
+Overview
+========
+
+This tutorial shows how to create your own simple `ROS plugin <http://wiki.ros.org/pluginlib>`_ for Costmap2D.
+
+Before starting the tutorial, please check this `video <https://vimeo.com/106994708>`_ which contains information about Costmap2D layers design and plugins basic operational principals.
+
+Requirements
+============
+
+It is assumed that ROS2, Gazebo and TurtleBot3 packages are installed or built locally. Please make sure that Navigation2 project is also built locally as it was made in :ref:`build-instructions`.
+
+Tutorial Steps
+==============
+
+1- Write a new Costmap2D plugin
+-------------------------------
+
+For better demonstration an example plugin will be used. It could be found in `navigation2_tutorials <https://github.com/ros-planning/navigation2_tutorials>`_ repository as ``nav2_gradient_costmap_plugin`` ROS2-package. This example plugin puts repeating costs gradients to the costmap. Please refer to it when making your own layer plugin for Costmap2D. Below is a brief description of most important parts of plugin's code.
+
+The plugin's main class ``nav2_gradient_costmap_plugin::GradientLayer`` is inherited from basic class ``nav2_costmap_2d::Layer``:
+
+.. code-block:: c
+
+  namespace nav2_gradient_costmap_plugin
+  {
+  
+  class GradientLayer : public nav2_costmap_2d::Layer
+
+The basic class provides the set of virtual methods API for working with costmap layers in a plugin. These methods are calling in runtime by ``LayeredCostmap``. The list of methods, their description and necessity to have these methods in plugin's code is presented in the table below:
+
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| **Virtual method**   | **Method description**                                                     | **Requires override?**  |
++----------------------+----------------------------------------------------------------------------+-------------------------+
+| onInitialize()       | Method is called at the end of plugin initialization. There is usually     | No                      |
+|                      | declarations of ROS parameters with its default values.                    |                         |
```

Comment:
Remove "with its default values" since it should be also getting the parameters if needed. Add "This is where any required initialization should occur."

Reply:
Done.